### PR TITLE
Render clocks dynamically instead of using images.

### PR DESCRIPTION
**Pull request by Lunar-Dawn**
This changes the rendering of the clocks from a simple image to a dynamic one using a CSS background and dynamic spokes. This has several advantages over the image-based approach:

* Clock sizes can be set to an arbitrary number (although the dialogue does limit it to 2-32).
* It allows the clocks to scale seamlessly to larger or smaller sizes if needed.
* The colours could be changed more easily, even potentially by users with a setting.

I am quite unfamiliar with both Foundy module development and handlebars, so I might have done something more subtle wrong as far as those are concerned. Apologies if so.

I have not tested your deployment pipeline since I don't quite know how it works, but the image directory export will probably need to be removed. Size restrictions are also only on the front end so `database.js` might want to enforce them too, but I wanted to make sure you were interested in the PR before I go beyond proof-of-concept.

The current look in Foundry:
image

### CODE REVIEW

**Comment by CarlosFdez** [comment]
Rendering dynamically is indeed something I've been wanting to do and only haven't done in the interest in time. I'll take a look soon, once I get more progress on stuff I'm working on.

**Comment by CarlosFdez** [comment]
So I started digging into it. Here are my general thoughts.
1) Avoid registering new handlebars helpers, this may interfere with certain system implementations.
2) Avoid let if at all possible. Try an IIFE if need be.
3) Are there system that go past 16 clock? The most I've ever seen was 12.

To handle the lack of a times helper, you could consider adding spokes to the render data. `spokes: Array(count).fill(0)` will handle that. If you wanna attach any additional data to each spoke render pass you could even do a map().

Past that, I have some more vague concerns about using a number input instead of a select, since a select is generally quicker in the heat of play. I can see select not being tenable when there are many options though. Do people actually make 11 clocks? Its easier to convince me of this though.

**Comment by CarlosFdez** [comment]
The general reason for limiting to 16 was that we could maybe keep it as a select. If its a number input there is no reason to limit it.

My current thought right now for a hybrid approach is actually a dropdown autocomplete with common sizes. I'll look into that in the coming days after I get some work done, but if you wanna take a peek beforehand you can.

Thanks for this, btw.

**Comment by Lunar-Dawn** [comment]
That does sound like a nice design with the dropdown. I suspect that getting it to nicely break out of the dialogue window might be somewhat difficult, but I'll take a shot at it.

No problem, I've wanted to get into Foundry module development for a while and this was a fun opportunity to try it.

**Comment by CarlosFdez** [comment]
Sounds good then. As for the dropdown, this might work if you can find a way to get the autocomplete to always show even if you just click it: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/datalist

ContextMenu unfortunately does have the problem you mentioned where it cannot exit the dialog. Any proper auto-complete needs to add to the document body and be absolutely positioned.

**Comment by Lunar-Dawn** [comment]
I tried `datalist`s first, but you cannot disable the filtering by current input, and the "show on focus" functionality is a bit testy and browser-dependent, so I wrote a custom solution instead.

The breakout from `overflow: hidden` is pretty clean, but it does mean that the dropdown size can't be defined in percentages in CSS, so it is currently just `100px`. It could be easily set with JS in the rendering hook but it does feel a bit dirty since none of the other rendering uses JS.

image

**Comment by Lunar-Dawn** [comment]
Tried it and there needs to be some limit on the clock sizes. 541adcd50879467881e0eb7d80169d4926a5786b sets a limit of 256 in the database, which I hope is the correct way to do it.

Otherwise the browser might try to render a few thousand spokes and hang (which is one of the reasons I considered writing the dynamic clock as `canvas`es instead before doing it with HTML/CSS).

**Comment by CarlosFdez** [comment]
I don't know how I didn't see these notifications. I'll check it out soon!

File: module/database.js [hunk 1]
```diff
@@ -20,6 +23,9 @@ export class ClockDatabase extends Collection {
     }
 
     update(data) {
+        if(!this.#verifyClockData(data))
+            return;
```
Comment (CarlosFdez) [inline module/database.js:27]:
```suggestion
        if (!this.#verifyClockData(data)) return;
```

File: module/database.js [hunk 2]
```diff
@@ -6,6 +6,9 @@
  */
 export class ClockDatabase extends Collection {
     addClock(data={}) {
+        if(!this.#verifyClockData(data))
+            return;
```
Comment (CarlosFdez) [inline module/database.js:10]:
```suggestion
        if (!this.#verifyClockData(data)) return;
```

**Review by CarlosFdez — COMMENTED** [review]
We should consider making a dialog subclass for the add/edit clock dialog now. (edit: I'll do this!)

File: module/clock-panel.js [hunk 3]
```diff
@@ -78,7 +87,8 @@ export class ClockPanel extends Application {
 
         $html.find("[data-action=add-clock]").on("click", async () => {
             const content = await renderTemplate("modules/global-progress-clocks/templates/clock-add-dialog.hbs", {
-                sizes: CLOCK_SIZES
+                max_size: CLOCK_MAX_SIZE,
+                preset_sizes: CLOCK_SIZES,
```
Comment (CarlosFdez) [inline module/clock-panel.js:91]:
```suggestion
                maxSize: CLOCK_MAX_SIZE,
                presetSizes: CLOCK_SIZES,
```

File: module/clock-panel.js [hunk 4]
```diff
@@ -100,7 +111,8 @@ export class ClockPanel extends Application {
 
             const content = await renderTemplate("modules/global-progress-clocks/templates/clock-add-dialog.hbs", {
                 clock,
-                sizes: CLOCK_SIZES,
+                max_size: CLOCK_MAX_SIZE,
+                preset_sizes: CLOCK_SIZES,
```
Comment (CarlosFdez) [inline module/clock-panel.js:115]:
```suggestion
                maxSize: CLOCK_MAX_SIZE,
                presetSizes: CLOCK_SIZES,
```

File: module/database.js [hunk 5]
```diff
@@ -54,4 +60,14 @@ export class ClockDatabase extends Collection {
...
     }
+
+    // Limit the clock max size to 256
+    // That's still way to high, but it's just an implementation check as to not hang/error out on enormous values
+    #verifyClockData(data) {
+        if(data.max < 256) {
+            return true;
+        }
+        ui.notifications.error(game.i18n.format("GlobalProgressClocks.SizeTooBigError", { max_size: 256 }));
+        return false;
```
Comment (CarlosFdez) [inline module/database.js:71]:
```suggestion
        if (data.max >= 256) {
            ui.notifications.error(game.i18n.format("GlobalProgressClocks.SizeTooBigError", { max_size: 256 }));
            return false;
        }
        
        return true;
```

File: styles/style.css [hunk 6]
```diff
@@ -83,9 +83,107 @@
...
+.add-clock-form .dropdown-wrapper input {
+    width: 100%;
+    height: 100%;
+
+    border: 0;
+    padding: 0;
+    margin: 0;
+
+    background: none;
+
```
Comment (CarlosFdez) [inline styles/style.css:146]:
```suggestion
```

File: templates/clock-add-dialog.hbs [hunk 7]
```diff
@@ -1,16 +1,22 @@
...
-        <select data-dtype="Number" name="max">
-            {{#select clock.max}}
-                {{#each sizes as |size|}}
-                    <option value="{{size}}">{{size}}</option>
+
+        <div class="dropdown-wrapper">
+            {{#if clock.max}}
+                {{ numberInput clock.max name="max" step=1 min=2 max=max_size }}
+            {{else}}
+                {{ numberInput 4         name="max" step=1 min=2 max=max_size }}
```
Comment (CarlosFdez) [inline templates/clock-add-dialog.hbs:13]:
```suggestion
                {{ numberInput 4 name="max" step=1 min=2 max=max_size }}
```

File: templates/clock-add-dialog.hbs [hunk 8]
```diff
@@ -1,16 +1,22 @@
...
+            {{/if}}
+            <ul class="dropdown">
+                {{#each preset_sizes as |size|}}
```
Comment (CarlosFdez) [inline templates/clock-add-dialog.hbs:16]:
```suggestion
                {{#each presetSizes as |size|}}
```

**Comment by CarlosFdez** [comment]
I moved the dialog to a new file. For my own personal curiousity, why is the dropdown not being clipped by parents having overflow:hidden? Normally even position absolute things have to deal with that.

**Comment by Lunar-Dawn** [comment]
As far as I understand it (although note that I haven't checked the specs for it), it is because absolute elements get rendered as if they were children of the closest non-`position: fixed` ancestor but get their default position from their "real" position in the tree.

The dropdown thus gets positioned relative to the wrapper but is rendered as if it were a child of the dialogue, which is outside the `overflow: hidden`. This can be seen if you set the `top` property or the dropdown width in percent, as those will be relative to the entire dialogue.

**Comment by CarlosFdez** [comment]
I see. top/left/right/bottom places it relative to the nearest relative, but without those it does try to put it in its current location and you can use margin to fake it. I guess that's the trick, as awkward as it is.

**Pull request merged**